packages/core/test/SignUp.ts

```diff
@@ -80,5 +81,150 @@ describe('Signup', function () {
+        })
+
+        it('should fail if the contract call was not called by admin', async () => {
+            const epoch = Number(await unirepContract.currentEpoch())
+            const epkNonce = 1
+            const randomEpochKey = genEpochKey(
+                genRandomSalt(),
+                epoch,
+                epkNonce
+            ).valueOf()
+            const oldUsername = genRandomSalt().valueOf()
+            const newUsername = genRandomSalt().valueOf()
+            const accounts = await ethers.getSigners()
+
+            await expect(
+                unirepSocialContract
+                    .connect(accounts[1])
+                    .setUsername(randomEpochKey, oldUsername, newUsername, {
+                        value: DEFAULT_ATTESTING_FEE,
+                    })
+            ).to.be.revertedWith(
+                'Only admin can send transactions to this contract'
+            )
+        })
+
+        it('should fail if a username is double registered', async () => {
+            const epoch = Number(await unirepContract.currentEpoch())
+            const epkNonce = 1
+            const randomEpochKey1 = genEpochKey(
+                genRandomSalt(),
+                epoch,
+                epkNonce
+            ).valueOf()
+
+            const oldUsername2 = genRandomSalt().valueOf()
+            const newUsername2 = genRandomSalt().valueOf()
+            const accounts = await ethers.getSigners()
+
+            // set newUserName to randomEpochKey1
+            await unirepSocialContract
+                .connect(accounts[0])
+                .setUsername(randomEpochKey1, oldUsername2, newUsername2, {
+                    value: DEFAULT_ATTESTING_FEE,
+                })
+                .then((t) => t.wait())
+
+            const oldUsername3 = 0
+            const randomEpochKey2 = genEpochKey(
+                genRandomSalt(),
+                epoch,
+                epkNonce
+            ).valueOf()
+
+            // try to set the same newUserName to randomEpochKey2
+            await expect(
+                unirepSocialContract
+                    .connect(accounts[0])
+                    .setUsername(randomEpochKey2, oldUsername3, newUsername2, {
+                        value: DEFAULT_ATTESTING_FEE,
+                    })
+            ).to.be.revertedWith('This username is already taken')
+        })
+
+        it('should mark old username as not claimed', async () => {
+            const epoch = Number(await unirepContract.currentEpoch())
+            const epkNonce = 1
+            const randomEpochKey = genEpochKey(
+                genRandomSalt(),
+                epoch,
+                epkNonce
+            ).valueOf()
+            const oldUsername4 = genRandomSalt().valueOf()
+            const newUsername4 = genRandomSalt().valueOf()
+            const accounts = await ethers.getSigners()
+
+            const tx = await unirepSocialContract
```

Comment:
Here you should claim the username before changing it. Otherwise the default value of the old username is 0 and we’re not testing anything